modules/netaddress.go

```diff
@@ -26,27 +34,87 @@ func (na NetAddress) Port() string {
 	if err != nil {
 		return ""
 	}
+	if na.IsValid() != nil {
+		return ""
+	}
 	return port
 }
 
-// IsLoopback returns true for ip addresses that are on the same machine.
-func (na NetAddress) IsLoopback() bool {
-	if !na.IsValid() {
+// isLoopback returns true for ip addresses that are on the same machine.
+func (na NetAddress) isLoopback() bool {
+	host, _, err := net.SplitHostPort(string(na))
+	if err != nil {
 		return false
 	}
-
-	host := na.Host()
-	if ip := net.ParseIP(host); ip != nil && ip.IsLoopback() {
+	if host == "localhost" {
 		return true
 	}
-	if host == "localhost" || host == "::" || host == "0.0.0.0" {
+	if ip := net.ParseIP(host); ip != nil && ip.IsLoopback() {
 		return true
 	}
 	return false
 }
 
-// IsValid does nothing. Please ignore it.
-func (na NetAddress) IsValid() bool {
-	_, _, err := net.SplitHostPort(string(na))
-	return err == nil
+// IsValid returns an error if the NetAddress is invalid. A valid NetAddress
+// is of the form "host:port", such that "host" is either a valid IPv4/IPv6
+// address or a valid hostname, and "port" is an integer in the range
+// [1,65535]. Furthermore, "host" may not be a loopback address (except during
+// testing). Valid IPv4 addresses, IPv6 addresses, and hostnames are detailed
+// in RFCs 791, 2460, and 952, respectively.
+func (na NetAddress) IsValid() error {
+	host, port, err := net.SplitHostPort(string(na))
+	if err != nil {
+		return err
+	}
+
+	portInt, err := strconv.Atoi(port)
+	if err != nil {
+		return errors.New("port is not an integer")
+	} else if portInt < 1 || portInt > 65535 {
+		return errors.New("port is invalid")
+	}
+
+	// This check must come after the valid port check so that a host such as
+	// "localhost:badport" will fail.
+	if na.isLoopback() {
+		if build.Release == "testing" {
+			return nil
+		} else {
+			return errors.New("host is a loopback address")
+		}
+	}
+
+	// First try to parse host as an IP address; if that fails, assume it is a
+	// hostname.
+	if ip := net.ParseIP(host); ip != nil {
+		if ip.IsUnspecified() {
+			return errors.New("host is the unspecified address")
+		}
+	} else {
+		if len(host) < 1 || len(host) > 253 {
+			return errors.New("invalid hostname length")
+		}
+		if host[0] == '-' {
+			return errors.New("hostnames cannot start with a hyphen")
+		}
```

Comment:
as per the RFC, labels can't begin *or end* with a hyphen. I would use `if strings.HasPrefix(host, "-") || strings.HasSuffix(host, "-")`